File: src/app/services/relayClient.ts
export interface RelayConfig {
  relayUrl: string;
  hostId: string;
  hostPublicKey: string;
}

type RelayConfigCandidate = {
  relayUrl?: string | null;
  hostId?: string | null;
  hostPublicKey?: string | null;
};

export interface RelayHttpRequest {
  method: string;
  path: string;
  headers?: Record<string, string>;
  body?: string | ArrayBuffer | Uint8Array | null;
}

export interface RelayHttpResponse {
  status: number;
  headers: Record<string, string>;
  data: unknown;
}

type RelayEventHandler = (message: { type: string; data?: unknown }) => void;

const encoder = new TextEncoder();
const decoder = new TextDecoder();
const PROTOCOL_LABEL = encoder.encode('clawchat-relay-v1');

function encodeBase64(bytes: Uint8Array): string {
  let binary = '';
  for (const byte of bytes) binary += String.fromCharCode(byte);
  return btoa(binary).replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/, '');
}

function decodeBase64(value: string): Uint8Array {
  const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
  const binary = atob(normalized + '='.repeat((4 - (normalized.length % 4)) % 4));
  return Uint8Array.from(binary, (char) => char.charCodeAt(0));
}

function websocketUrl(relayUrl: string, hostId: string): string {
  const base = relayUrl
    .replace(/^http:/, 'ws:')
    .replace(/^https:/, 'wss:')
    .replace(/\/+$/, '');
  return `${base}/v1/relay/client/${encodeURIComponent(hostId)}`;
}

function requestBytes(body: RelayHttpRequest['body']): Uint8Array {
  if (body == null) return new Uint8Array();
  if (typeof body === 'string') return encoder.encode(body);
  if (body instanceof Uint8Array) return body;
  return new Uint8Array(body);
}

function asArrayBuffer(bytes: Uint8Array): ArrayBuffer {
  return bytes.buffer.slice(bytes.byteOffset, bytes.byteOffset + bytes.byteLength) as ArrayBuffer;
}

class RelayClient {
  private socket: WebSocket | null = null;
  private config: RelayConfig | null = null;
  private encryptionKey: CryptoKey | null = null;
  private connectPromise: Promise<void> | null = null;
  private resolveConnect: (() => void) | null = null;
  private rejectConnect: ((error: Error) => void) | null = null;
  private pending = new Map<
    string,
    { resolve: (value: RelayHttpResponse) => void; reject: (error: Error) => void }
  >();
  private eventHandlers = new Set<RelayEventHandler>();
  private requestCounter = 0;
  private receivedNonces = new Set<string>();
  private nonceOrder: string[] = [];

  isConfigured(config: RelayConfigCandidate): config is RelayConfig {
    return Boolean(config.relayUrl && config.hostId && config.hostPublicKey);
  }

  async connect(config: RelayConfig): Promise<void> {
    if (this.socket?.readyState === WebSocket.OPEN && this.encryptionKey && this.sameConfig(config))
      return;
    if (this.connectPromise && this.sameConfig(config)) return this.connectPromise;

    this.disconnect();
    this.config = config;
    const connection = new Promise<void>((resolve, reject) => {
      this.resolveConnect = resolve;
      this.rejectConnect = reject;
    });
    this.connectPromise = connection;

    try {
      const keyPair = (await crypto.subtle.generateKey({ name: 'X25519' }, true, [
        'deriveBits',
      ])) as CryptoKeyPair;
      const hostKey = await crypto.subtle.importKey(
        'raw',
        asArrayBuffer(decodeBase64(config.hostPublicKey)),
        { name: 'X25519' },
        false,
        [],
      );
      const sharedSecret = await crypto.subtle.deriveBits(
        { name: 'X25519', public: hostKey },
        keyPair.privateKey,
        256,
      );
      const hkdfKey = await crypto.subtle.importKey('raw', sharedSecret, 'HKDF', false, [
        'deriveKey',
      ]);
      this.encryptionKey = await crypto.subtle.deriveKey(
        {
          name: 'HKDF',
          hash: 'SHA-256',
          salt: encoder.encode(config.hostId),
          info: PROTOCOL_LABEL,
        },
        hkdfKey,
        { name: 'AES-GCM', length: 256 },
        false,
        ['encrypt', 'decrypt'],
      );
      const publicKey = new Uint8Array(await crypto.subtle.exportKey('raw', keyPair.publicKey));
      const socket = new WebSocket(websocketUrl(config.relayUrl, config.hostId));
      this.socket = socket;
      socket.onopen = () =>
        socket.send(
          JSON.stringify({
            kind: 'hello',
            client_public_key: encodeBase64(publicKey),
          }),
        );
      socket.onmessage = (event) => {
        void this.handleFrame(event.data);
      };
      socket.onerror = () => {
        if (this.socket === socket) this.failConnection(new Error('Relay connection failed'));
      };
      socket.onclose = () => {
        if (this.socket === socket) this.failConnection(new Error('Relay connection closed'));
      };
    } catch (error) {
      this.failConnection(error instanceof Error ? error : new Error('Relay setup failed'));
    }
    return connection;
  }

  disconnect(): void {
    const socket = this.socket;
    this.socket = null;
    if (socket && socket.readyState < WebSocket.CLOSING) socket.close(1000, 'Client disconnect');
    this.encryptionKey = null;
    this.connectPromise = null;
    this.resolveConnect = null;
    this.rejectConnect = null;
    for (const pending of this.pending.values()) pending.reject(new Error('Relay disconnected'));
    this.pending.clear();
    this.receivedNonces.clear();
    this.nonceOrder = [];
  }

  onEvent(handler: RelayEventHandler): () => void {
    this.eventHandlers.add(handler);
    return () => this.eventHandlers.delete(handler);
  }

  async subscribe(config: RelayConfig, token: string): Promise<void> {
    await this.connect(config);
    await this.sendEncrypted({ type: 'subscribe', token });
  }

  async unsubscribe(): Promise<void> {
    if (this.socket?.readyState === WebSocket.OPEN && this.encryptionKey) {
      await this.sendEncrypted({ type: 'unsubscribe' });
    }
  }

  async request(config: RelayConfig, request: RelayHttpRequest): Promise<RelayHttpResponse> {
    await this.connect(config);
    const id = `relay-${Date.now()}-${++this.requestCounter}`;
    const response = new Promise<RelayHttpResponse>((resolve, reject) => {
      const timeout = setTimeout(() => {
        this.pending.delete(id);
        reject(new Error('Relay request timed out'));
      }, 120_000);
      this.pending.set(id, {
        resolve: (value) => {
          clearTimeout(timeout);
          resolve(value);
        },
        reject: (error) => {
          clearTimeout(timeout);
          reject(error);
        },
      });
    });
    await this.sendEncrypted({
      type: 'http_request',
      id,
      method: request.method,
      path: request.path,
      headers: request.headers ?? {},
      body: encodeBase64(requestBytes(request.body)),
    });
    return response;
  }

  private async sendEncrypted(payload: object): Promise<void> {
    if (!this.encryptionKey || this.socket?.readyState !== WebSocket.OPEN) {
      throw new Error('Relay is not connected');
    }
    const nonce = crypto.getRandomValues(new Uint8Array(12));
    const ciphertext = await crypto.subtle.encrypt(
      { name: 'AES-GCM', iv: nonce, additionalData: PROTOCOL_LABEL },
      this.encryptionKey,
      encoder.encode(JSON.stringify(payload)),
    );
    this.socket.send(
      JSON.stringify({
        kind: 'encrypted',
        nonce: encodeBase64(nonce),
        ciphertext: encodeBase64(new Uint8Array(ciphertext)),
      }),
    );
  }

  private async handleFrame(raw: string): Promise<void> {
    const frame = JSON.parse(raw);
    if (frame.kind === 'host_offline') {
      this.failConnection(new Error('ClawChat host is offline'));
      return;
    }
    if (frame.kind !== 'encrypted' || !this.encryptionKey) return;
    try {
      if (this.receivedNonces.has(frame.nonce)) throw new Error('Relay frame was replayed');
      const plaintext = await crypto.subtle.decrypt(
        {
          name: 'AES-GCM',
          iv: asArrayBuffer(decodeBase64(frame.nonce)),
          additionalData: PROTOCOL_LABEL,
        },
        this.encryptionKey,
        asArrayBuffer(decodeBase64(frame.ciphertext)),
      );
      const payload = JSON.parse(decoder.decode(plaintext));
      this.receivedNonces.add(frame.nonce);
      this.nonceOrder.push(frame.nonce);
      if (this.nonceOrder.length > 2048) {
        this.receivedNonces.delete(this.nonceOrder.shift()!);
      }
      if (payload.type === 'ready') {
        if (payload.host_id !== this.config?.hostId)
          throw new Error('Relay host identity mismatch');
        this.resolveConnect?.();
        this.resolveConnect = null;
        this.rejectConnect = null;
      } else if (payload.type === 'http_response') {
        const pending = this.pending.get(payload.id);
        if (!pending) return;
        this.pending.delete(payload.id);
        const bytes = decodeBase64(payload.body ?? '');
        const contentType = payload.headers?.['content-type'] ?? '';
        let data: unknown = bytes;
        if (contentType.includes('application/json')) {
          data = bytes.length ? JSON.parse(decoder.decode(bytes)) : null;
        } else if (contentType.startsWith('text/')) {
          data = decoder.decode(bytes);
        }
        pending.resolve({ status: payload.status, headers: payload.headers ?? {}, data });
      } else if (payload.type === 'event') {
        for (const handler of this.eventHandlers) handler(payload.data);
      } else if (payload.type === 'auth_error') {
        for (const handler of this.eventHandlers) handler({ type: 'auth_error' });
      }
    } catch (error) {
      this.failConnection(error instanceof Error ? error : new Error('Invalid relay frame'));
    }
  }

  private sameConfig(config: RelayConfig): boolean {
    return (
      this.config?.relayUrl === config.relayUrl &&
      this.config.hostId === config.hostId &&
      this.config.hostPublicKey === config.hostPublicKey
    );
  }

  private failConnection(error: Error): void {
    const socket = this.socket;
    this.socket = null;
    if (socket && socket.readyState < WebSocket.CLOSING) socket.close();
    this.encryptionKey = null;
    this.rejectConnect?.(error);
    this.rejectConnect = null;
    this.resolveConnect = null;
    this.connectPromise = null;
    for (const pending of this.pending.values()) pending.reject(error);
    this.pending.clear();
    for (const handler of this.eventHandlers) handler({ type: 'relay_disconnected' });
  }
}

export const relayClient = new RelayClient();

File: android/core/src/main/java/com/clawchat/android/core/network/RelayClient.kt
package com.clawchat.android.core.network

import android.util.Base64
import android.util.Log
import com.clawchat.android.core.data.ActiveSession
import com.clawchat.android.core.data.SessionStore
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withTimeout
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Protocol
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody.Companion.toResponseBody
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.Buffer
import org.bouncycastle.crypto.agreement.X25519Agreement
import org.bouncycastle.crypto.digests.SHA256Digest
import org.bouncycastle.crypto.generators.HKDFBytesGenerator
import org.bouncycastle.crypto.params.HKDFParameters
import org.bouncycastle.crypto.params.X25519PrivateKeyParameters
import org.bouncycastle.crypto.params.X25519PublicKeyParameters
import org.json.JSONObject
import java.security.SecureRandom
import java.util.UUID
import java.util.ArrayDeque
import java.util.concurrent.ConcurrentHashMap
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec
import javax.inject.Inject
import javax.inject.Singleton

data class RelayResponse(val status: Int, val contentType: String?, val body: ByteArray)

@Singleton
class RelayClient @Inject constructor(private val sessionStore: SessionStore) {
    companion object {
        private const val TAG = "RelayClient"
        private val LABEL = "clawchat-relay-v1".toByteArray()
    }

    private val httpClient = OkHttpClient()
    private val connectionMutex = Mutex()
    private val pending = ConcurrentHashMap<String, CompletableDeferred<RelayResponse>>()
    private val receivedNonces = mutableSetOf<String>()
    private val nonceOrder = ArrayDeque<String>()
    private val _events = MutableSharedFlow<String>(extraBufferCapacity = 64)
    val events: SharedFlow<String> = _events.asSharedFlow()

    @Volatile private var webSocket: WebSocket? = null
    @Volatile private var aesKey: ByteArray? = null
    @Volatile private var ready: CompletableDeferred<Unit>? = null
    @Volatile private var activeHostId: String? = null

    suspend fun ensureConnected(expectedScope: String? = null): Boolean = connectionMutex.withLock {
        val session = sessionStore.activeSession.first()
        if (expectedScope != null && session?.connectionScope() != expectedScope) {
            throw SessionScopeChangedException(expectedScope, session?.connectionScope())
        }
        // Relay pairing is host-scoped. A manual direct session can retain old
        // pairing preferences during a transition, but share work must never
        // fall back through that unrelated host.
        if (expectedScope != null && session?.authMode != "paired") {
            throw SessionScopeChangedException(expectedScope, session?.connectionScope())
        }
        val relayUrl = sessionStore.relayUrl.first() ?: return false
        val hostId = sessionStore.hostId.first() ?: return false
        val hostPublicKey = sessionStore.hostPublicKey.first() ?: return false
        if (expectedScope != null && hostId != expectedScope) {
            throw SessionScopeChangedException(expectedScope, hostId)
        }
        val latestSession = sessionStore.activeSession.first()
        if (expectedScope != null && latestSession?.connectionScope() != expectedScope) {
            throw SessionScopeChangedException(expectedScope, latestSession?.connectionScope())
        }
        if (activeHostId == hostId && ready?.isCompleted == true && webSocket != null) return true

        disconnectInternal()
        val privateKey = X25519PrivateKeyParameters(SecureRandom())
        val publicKey = privateKey.generatePublicKey().encoded
        val agreement = X25519Agreement().apply { init(privateKey) }
        val sharedSecret = ByteArray(agreement.agreementSize)
        agreement.calculateAgreement(X25519PublicKeyParameters(decode(hostPublicKey), 0), sharedSecret, 0)
        aesKey = ByteArray(32).also { derivedKey ->
            HKDFBytesGenerator(SHA256Digest()).apply {
                init(HKDFParameters(sharedSecret, hostId.toByteArray(), LABEL))
                generateBytes(derivedKey, 0, derivedKey.size)
            }
        }
        activeHostId = hostId
        val readySignal = CompletableDeferred<Unit>()
        ready = readySignal
        val wsUrl = relayUrl.replaceFirst("https://", "wss://")
            .replaceFirst("http://", "ws://").trimEnd('/') + "/v1/relay/client/$hostId"
        webSocket = httpClient.newWebSocket(Request.Builder().url(wsUrl).build(), object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                webSocket.send(JSONObject().apply {
                    put("kind", "hello")
                    put("client_public_key", encode(publicKey))
                }.toString())
            }

            override fun onMessage(webSocket: WebSocket, text: String) = handleFrame(text)
            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                if (this@RelayClient.webSocket === webSocket) failConnection(t)
            }
            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                if (this@RelayClient.webSocket === webSocket) {
                    failConnection(IllegalStateException("Relay closed: $reason"))
                }
            }
        })
        try {
            withTimeout(15_000) { readySignal.await() }
            true
        } catch (error: Exception) {
            Log.w(TAG, "Relay handshake failed", error)
            disconnectInternal()
            false
        }
    }

    suspend fun subscribe(token: String): Boolean {
        if (!ensureConnected()) return false
        sendEncrypted(JSONObject().put("type", "subscribe").put("token", token))
        return true
    }

    suspend fun unsubscribe() {
        if (ready?.isCompleted == true) runCatching {
            sendEncrypted(JSONObject().put("type", "unsubscribe"))
        }
    }

    suspend fun execute(request: Request): Response {
        val expectedScope = request.tag(ExpectedSessionScope::class.java)?.value
        if (!ensureConnected(expectedScope)) throw IllegalStateException("Relay is not configured")
        val id = UUID.randomUUID().toString()
        val deferred = CompletableDeferred<RelayResponse>()
        pending[id] = deferred
        val bodyBytes = request.body?.let { body ->
            Buffer().use { buffer -> body.writeTo(buffer); buffer.readByteArray() }
        } ?: ByteArray(0)
        val path = request.url.encodedPath + request.url.encodedQuery?.let { "?$it" }.orEmpty()
        sendEncrypted(JSONObject().apply {
            put("type", "http_request")
            put("id", id)
            put("method", request.method)
            put("path", path)
            put("headers", JSONObject().apply {
                request.header("Authorization")?.let { put("authorization", it) }
                request.header("Content-Type")?.let { put("content-type", it) }
                request.header("Accept")?.let { put("accept", it) }
            })
            put("body", encode(bodyBytes))
        })
        val relayResponse = try {
            withTimeout(120_000) { deferred.await() }
        } finally {
            pending.remove(id)
        }
        return Response.Builder().request(request).protocol(Protocol.HTTP_1_1)
            .code(relayResponse.status).message("Relay")
            .body(relayResponse.body.toResponseBody(relayResponse.contentType?.toMediaTypeOrNull()))
            .build()
    }

    private fun sendEncrypted(payload: JSONObject) {
        val key = aesKey ?: error("Relay key is unavailable")
        val nonce = ByteArray(12).also(SecureRandom()::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(key, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(LABEL)
        val ciphertext = cipher.doFinal(payload.toString().toByteArray())
        check(webSocket?.send(JSONObject().apply {
            put("kind", "encrypted")
            put("nonce", encode(nonce))
            put("ciphertext", encode(ciphertext))
        }.toString()) == true) { "Relay send failed" }
    }

    private fun handleFrame(text: String) {
        try {
            val frame = JSONObject(text)
            if (frame.optString("kind") == "host_offline") error("Host is offline")
            if (frame.optString("kind") != "encrypted") return
            val nonceValue = frame.getString("nonce")
            synchronized(receivedNonces) {
                check(nonceValue !in receivedNonces) { "Relay frame was replayed" }
            }
            val cipher = Cipher.getInstance("AES/GCM/NoPadding")
            cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(aesKey ?: return, "AES"),
                GCMParameterSpec(128, decode(nonceValue)))
            cipher.updateAAD(LABEL)
            val payload = JSONObject(String(cipher.doFinal(decode(frame.getString("ciphertext")))))
            synchronized(receivedNonces) {
                receivedNonces.add(nonceValue)
                nonceOrder.addLast(nonceValue)
                if (nonceOrder.size > 2048) receivedNonces.remove(nonceOrder.removeFirst())
            }
            when (payload.optString("type")) {
                "ready" -> {
                    check(payload.getString("host_id") == activeHostId) { "Relay host mismatch" }
                    ready?.complete(Unit)
                }
                "http_response" -> pending[payload.getString("id")]?.complete(RelayResponse(
                    payload.getInt("status"),
                    payload.optJSONObject("headers")?.optString("content-type"),
                    decode(payload.optString("body")),
                ))
                "event" -> payload.optJSONObject("data")?.let { _events.tryEmit(it.toString()) }
                "auth_error" -> _events.tryEmit("""{"type":"auth_error"}""")
            }
        } catch (error: Exception) {
            failConnection(error)
        }
    }

    private fun failConnection(error: Throwable) {
        ready?.completeExceptionally(error)
        pending.values.forEach { it.completeExceptionally(error) }
        pending.clear()
        _events.tryEmit("""{"type":"relay_disconnected"}""")
        webSocket = null
    }

    private fun disconnectInternal() {
        webSocket?.close(1000, "Reconnect")
        webSocket = null
        aesKey = null
        ready = null
        activeHostId = null
        synchronized(receivedNonces) {
            receivedNonces.clear()
            nonceOrder.clear()
        }
    }

    private fun encode(bytes: ByteArray): String =
        Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    private fun decode(value: String): ByteArray =
        Base64.decode(value, Base64.URL_SAFE or Base64.NO_WRAP or Base64.NO_PADDING)

    private fun ActiveSession.connectionScope(): String? = when (authMode) {
        "paired" -> hostId
        "manual" -> apiBaseUrl?.trimEnd('/')
        else -> hostId ?: apiBaseUrl?.trimEnd('/')
    }
}
